### ISSUE TITLE
OutOfMemoryException in MS.CA.Diagnostics.CompilationWithAnalyzers.ComputeAnalyzerDiagnosticsAsync

### ISSUE THREAD

**Issue by pakdev** [dotnet/roslyn#19661]
**Version Used**: 2.02.0.61624 (MS.CA)

**Steps to Reproduce**:

1. Load 14 projects in a solution containing 300+ in Visual Studio 2017 (rest disabled with [vsFunnel](http://vsfunnel.com/))
2. Edit files for ~2.5 hours

**StackTrace**:
```
   at System.Collections.Immutable.ImmutableArray`1.Builder.ToArray()
   at System.Collections.Immutable.ImmutableArray`1.Builder.ToImmutable()
   at Microsoft.CodeAnalysis.CSharp.Binder.AttributeExpressionVisitor.VisitArguments(ImmutableArray`1 arguments, DiagnosticBag diagnostics, Boolean& attrHasErrors, Boolean parentHasErrors)
   at Microsoft.CodeAnalysis.CSharp.Binder.GetAttribute(BoundAttribute boundAttribute, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.GetAttribute(AttributeSyntax node, NamedTypeSymbol boundAttributeType, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.GetAttributes(ImmutableArray`1 binders, ImmutableArray`1 attributesToBind, ImmutableArray`1 boundAttributeTypes, CSharpAttributeData[] attributesBuilder, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Symbol.LoadAndValidateAttributes(OneOrMany`1 attributesSyntaxLists, CustomAttributesBag`1& lazyCustomAttributesBag, AttributeLocation symbolPart, Boolean earlyDecodingOnly, DiagnosticBag addToDiagnostics, Binder binderOpt)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceAssemblySymbol.EnsureAttributesAreBound()
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceAssemblySymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetSourceDeclarationDiagnostics(SyntaxTree syntaxTree, Nullable`1 filterSpanWithinTree, Func`4 locationFilterOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetDiagnosticsForSyntaxTree(CompilationStage stage, SyntaxTree syntaxTree, Nullable`1 filterSpanWithinTree, Boolean includeEarlierStages, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetDiagnostics(Nullable`1 span, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.GenerateCompilationEvents(AnalysisScope analysisScope, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.<GenerateCompilationEventsAndPopulateEventsCacheAsync>d__57.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.<ComputeAnalyzerDiagnosticsAsync>d__56.MoveNext()
```

Notes:
- ReSharper is **not** installed.
- Crash dump lists 32 analyzers that appear to be built-ins e.g. IDE0031, CSxxxx, IDE0027, etc...
- I can share the 3GB+ crash dump

**Expected Behavior**: Does not crash

**Actual Behavior**: Crashed

**Comment by Pilchie** [comment]
@mavasani or @heejaechang care to take a look?

**Comment by heejaechang** [comment]
@pakdev did you reload your solution multiple times in same VS? and yes, I would like to take a look at the dump if possible. thank you for reporting!

**Comment by pakdev** [comment]
The solution was not reloaded multiple times. I've reported a problem in VS, referenced this issue, and attached the dump file. Please let me know if that won't suffice.

**Comment by heejaechang** [comment]
@pakdev are there .net core projects?

**Comment by pakdev** [comment]
@heejaechang Nope

**Comment by Pilchie** [comment]
Also tagging @sharwell who has been looking at some OOMs lately.

**Comment by MattGertz** [comment]
Mondo escrow triage:  meets the bar; definitely want to understand OOM for Escrow.

**Comment by heejaechang** [comment]
@pakdev can you share the dump? 

also, do you have code lens enabled?

**Comment by sharwell** [comment]
I haven't hit an OOM situation yet with 15.3 Preview 3. I checked with @CyrusNajmabadi and he hasn't either. We used to hit these multiple times daily, so while there is still work to do regarding allocation rates and performance we're feeling confident that we've (in conjunction with other teams) have tremendously improved stability with respect to OOM conditions in large projects.

**Comment by heejaechang** [comment]
another way to fix this issue once for all is that we swallow the bullet and move OOP to 64bit. (this is one line change, so good candidate for A/B testing?) @jinujoseph

**Comment by pakdev** [comment]
@heejaechang Is there a way I can securely transfer these dumps to you? 

@sharwell I'll have my guys try the Preview 3 (I assume you mean [this](https://www.visualstudio.com/en-us/news/releasenotes/vs2017-preview-relnotes) one)

**Comment by heejaechang** [comment]
@pakdev in VS. click help->Send Feedback -> Report problem -> click big (+) below -> in the description put my alias (hechang) -> next -> attach the dump -> next -> submit

**Comment by sharwell** [comment]
@pakdev That's the one! We've made some big improvements since then as well, but at this point the memory work is more about speed than reliability.

**Comment by pakdev** [comment]
@heejaechang Sent the dump (I have others if interested - through the OOM location(s) are different). Also, the developer says he never set the code lens option and the default appears to be "off".

**Comment by heejaechang** [comment]
@pakdev default is on for code lens.

**Comment by pakdev** [comment]
@heejaechang Hmm, then I can't answer that question definitively. Either the user had it disabled at the time of the crash or disabled it afterwards in an attempt to avoid crashes.

**Comment by heejaechang** [comment]
I believe we fixed most of OOM cases we know of in latest 15.4

please re-open this if it is still an issue.
